tool/planet/agent.go

```diff
@@ -143,85 +121,73 @@ func startLeaderClient(config agentConfig, agent agent.Agent, errorC chan error)
-		// recordEventTimeout specifies the max timeout to record an event
-		const recordEventTimeout = 10 * time.Second
+	// Watch for changes in the election mode.
+	// start/stop voter participation depending on the value of the election mode key.
+	client.AddWatchCallback(conf.ElectionKey, conf.Term, manageParticipation(client, conf, errorC))
 
-		if newVal != conf.PublicIP {
-			return
-		}
-		// Ignore if same leader is re-elected
-		if newVal == prevVal {
-			return
-		}
+	// Watch for changes to the leader key.
+	// If this agent becomes the leader, record a LeaderElected event to the local timeline.
+	client.AddWatchCallback(conf.LeaderKey, conf.Term/3, recordElectionEvents(conf.PublicIP, agent))
 
-		ctx, cancel := context.WithTimeout(context.Background(), recordEventTimeout)
-		defer cancel()
+	if !conf.HighAvailability {
+		// Watch for changes to the leader key.
+		// If this agent becomes the leader, start a number of additional services.
+		client.AddWatchCallback(conf.LeaderKey, conf.Term/3, manageUnits(config))
+	}
 
-		agent.RecordLocalEvents(ctx, []*pb.TimelineEvent{
-			pb.NewLeaderElected(agent.GetConfig().Clock.Now(), prevVal, newVal),
-		})
-	})
+	// Set initial value of election mode
+	etcdapi := etcd.NewKeysAPI(etcdClient)
+	_, err = etcdapi.Set(context.TODO(), conf.ElectionKey,
+		strconv.FormatBool(conf.ElectionEnabled),
+		&etcd.SetOptions{PrevExist: etcd.PrevNoExist})
+	if !trace.IsAlreadyExists(convertError(err)) {
+		return nil, trace.Wrap(err)
+	}
 
-	var cancelVoter context.CancelFunc
-	var ctx context.Context
-	if conf.Role == RoleMaster {
-		switch conf.ElectionEnabled {
-		case true:
-			log.Infof("Adding voter for IP %v.", conf.PublicIP)
-			ctx, cancelVoter = context.WithCancel(context.TODO())
-			if err = client.AddVoter(ctx, conf.LeaderKey, conf.PublicIP, conf.Term); err != nil {
-				return nil, trace.Wrap(err)
-			}
-		case false:
-			log.Info("Shut down services until election has been re-enabled.")
-			// Shut down services at startup if running as master
-			if err := stopUnits(context.TODO()); err != nil {
-				log.WithError(err).Warn("Failed to stop units.")
-			}
+	// Start control plane components on all master nodes if running in HA mode.
+	if conf.HighAvailability {
+		if err := startUnits(context.TODO()); err != nil {
+			return nil, trace.Wrap(err, "failed to start units")
+		}
+		if err := validateKubernetesService(context.TODO(), config.serviceCIDR); err != nil {
+			return nil, trace.Wrap(err, "failed to validate kubernetes services")
 		}
 	}
-	// watch the election mode status and start/stop participation
-	// depending on the value of the election mode key
-	client.AddWatchCallback(conf.ElectionKey, conf.Term, func(key, prevVal, newVal string) {
-		var err error
-		enabled, _ := strconv.ParseBool(newVal)
+
+	return client, nil
+}
+
+// manageParticipation starts voter participation when the election mode is
```

Comment:
nitpick: referring to this as election mode here and above reads a bit weird to me. I would consider rewording this a bit to just say whether elections are enabled/disabled.